### ISSUE TITLE
Feature Request: Use OSM data to set grid cell land use

### ISSUE THREAD

**Issue by Claudio9701** [CityScope/CS_cityscopeJS#110]
Hi! I was wondering if there is anyone currently working on this feature? I started to check if it is possible to do this on this [issue](https://github.com/CityScope/CS_Brix/issues/14#issuecomment-739553732), but it was the wrong repo. 

I have already played a bit with the idea but in Python on this [colab](https://colab.research.google.com/drive/1WjDXdYBoSsugDwVF0W5YdhtcN5D9UGj3?usp=sharing). Even though I know some javascript I felt more comfortable doing the functionality in Python before and then port it to js. I think js libraries that could help a lot are:

- Geodata downloading: query-overpass, overpass-frontend or osm2geojson-lite
- Geodata processing: turf & d3

Thanks for your time!

**Comment by RELNO** [comment]
Thanks @Claudio9701 -- it's a great feature and I'd love to think of it as soon as I've more time. 
In the meantime, if you want to take a stab:
https://github.com/CityScope/CS_cityscopeJS/blob/master/src/components/GridEditor/EditorMenu/GridProps/GridMaker/GridMaker.js is where one creats a random grid whcih is populated with the types in types table below.
This calls https://github.com/CityScope/CS_cityscopeJS/blob/master/src/components/GridEditor/EditorMenu/GridProps/GridMaker/gridCreator.js that actually creates a simple GEOjson grid and then dispatchs it to Redux via:

```
   const handleGridCreation = () => {
        let grid = gridCreator(props.gridProps);
        dispatch(listenToGridCreator(grid));
    };

```
So in practice, your code can be a simple function that seats right after the grid is created, and allow to replace the random values with the OSM query.

**Comment by Claudio9701** [comment]
Hi :wave:

I'm working on this [observablehq notebook](https://observablehq.com/d/91be7c93ca95ae73) to test the implementation. At this moment I have been able to query osm and do some spatial data processing (buffering and enveloping points). The next steps are:
- Assign a valid land-use to the osm data
- Calculate grid cell centroids
- Get land-use for each cell centroid

**Comment by RELNO** [comment]
@Claudio9701 looks like a great start! Keep in mind that in CSjs grid editor, most of the above would be already done by the tool (getting grid, finding grid points, etc). What the editor currently missing is your last  part, where you correlate land uses to grid cells. When you're done, we'll see how to built this directly into the editor.

**Comment by Claudio9701** [comment]
@RELNO I was able to finish the steps, the final result looks like this:

![image](https://user-images.githubusercontent.com/24416566/101977073-f6a7b600-3c18-11eb-9a12-4ee448a51aa7.png)

Nevertheless, I consider that is too "naive" to assign the default landuse Residential, maybe It would be better to get the road network, filter cells that intersect with it, and set landuse to none or something similar.

Also, I guess this approach could work better on places with more osm data, maybe Lima isn't a great example.

**Comment by RELNO** [comment]
@Claudio9701 looks great -- I think that the logic is clearly there, now it's a matter of building it into the editor app.
One question @crisjf  and I discussed is that the 'types' (land uses) in the editor app are created by the user, not pulled from OSM. I wonder how would we match the users' type lists to what we get from OSM.

**Comment by Claudio9701** [comment]
@RELNO thanks, yes I will try to implement this in the codebase and give you the updates.

Considering user-defined land uses may be out of the scope. I think of this feature as a helper to initialize the grid with land uses "best guess" using OSM data but further customization may require user intervention. Despite this, a solution can be implementing a form and use it to generate a custom `set_landuse_types()` function, but IMO this could be pretty hard to do and maintain because of the countless OSM tags available.

This would be easier using a proper Land Use/Cover API, as it would reduce the number of land uses types that a user can select. Have you thought about making one? I haven't found any useful resource that allows querying land uses using coordinates.

## TL;DR:
- Maybe custom land uses are too hard to implement, this could help the grid initialization and further customization would need user "manual" input.
- Using a Land Use/Cover API (I haven't found a useful one) can solve this problem, and reduce the number of land uses types that a user can select.

**Comment by crisjf** [comment]
@RELNO a solution would be that the user defines a map between their defined uses and the OSM tags. Much like the user now defines the map between the `types` and `NAICS` or `LBCS`. Enforce that mapping to be 1-1 and you're done.

**Comment by RELNO** [comment]
Both solutions are valid -- I'm personally leaning towards a solution that basically lets the use decide what set of land uses ('types') they would like to have. If the user clicks the 'get grid data from OSM', that grid will be populated with OSM vals. 
Otherwise, the grid will be random set of land uses from the user list (as it is today). In both cases, the user could overwrite OSM or random grid with the brush tool, so that the grid can be customized as needed.

**Comment by Claudio9701** [comment]
Hi! I've made some advancements with the implementation [here](https://github.com/Claudio9701/CS_cityscopeJS). 

![output](https://user-images.githubusercontent.com/24416566/103105457-f71e5480-45fb-11eb-913f-3969aaf62576.gif)

My main concerns with this are:

1. Assign an "intelligent" default land-use (catch roads, water sources, among others) (https://wiki.openstreetmap.org/wiki/Tag:place%3Dsea, https://wiki.openstreetmap.org/wiki/Tag:place%3Docean, https://gis.stackexchange.com/questions/341746/what-is-a-correct-overpass-turbo-query-for-getting-all-streets-in-a-city, )
1. Verify the correct use of async/await functions and if any redux-thunk middleware is needed (https://dmitripavlutin.com/javascript-fetch-async-await/, https://redux.js.org/tutorials/fundamentals/part-6-async-logic)
1. Review alternatives for generating interactively [high-resolution urban land-use maps](https://www.sciencedirect.com/science/article/abs/pii/S0034425718302074) with satellite imagery and osm data using some model

**Comment by RELNO** [comment]
@Claudio9701 very exciting! 
I could not get this work more then once (see the crash report below),  but this is very much along the lines of what we should have, great job! 
A few comments/ideas:
- I don't think we need to map the types to OSM; In other words, if the user selects 'get types from OSM', then a set of new types should be added to the table, stemming from OSM layer data. The user will then be able to use the 'brush' to edit types at will (for example, to create a future intervention that is not there yet). When the table is then sent to cityIO, both OSM types as well as user defined types will follow. 
- Why do we have to use Thunk? You can freeze the app (with a full screen dialog or snackbar https://material-ui.com/components/dialogs/) and wait till GET reqs are done. 
- I was trying to use the feature more then once on different locations, and it crashed:
<img width="968" alt="Screen Shot 2020-12-24 at 16 08 05" src="https://user-images.githubusercontent.com/14296774/103106465-c93d0e00-4603-11eb-963c-60169b00567a.png">

**Comment by Claudio9701** [comment]
Thanks for the quick answer!

- I think that would be easier. Just to be sure I'm getting this right, does the set of new types are the ones from the [OSM landuse key wikipage](https://wiki.openstreetmap.org/wiki/Key:landuse)? If this is the case is worth noting that there are [other OSM tags used for reporting landuse](https://wiki.openstreetmap.org/wiki/Land_use)
- Oh that's great, I will check how to use the materialui dialogs. :smile: 
- Sorry, I skipped implementing try/catch logic to handle empty responses from the Overpass API :facepalm:

**Comment by RELNO** [comment]
we need to find a way to convert OSM default landuse values (from whichever source) to CityScope types with LBCS params. I would go with the most common/used OSM landuse system (is this the one? https://wiki.openstreetmap.org/wiki/Key:landuse) and use a conversion method as shown here:  https://www.researchgate.net/publication/322089643_Quantitative_Comparison_of_Open-Source_Data_for_Fine-Grain_Mapping_of_Land_Use

**Comment by crisjf** [comment]
I've tried using the `landuse` tag before. Be aware that it is very sparse. I know that @doorleyr usually uses a combination of multiple tags. 

Here's an example:
```
amenity_tag_categories={
    "restaurants": {
      "amenity":["restaurant","cafe","fast_food","pub","cafe"],
      "shop":["coffee"]
    },
    "nightlife": {
      "amenity":["bar","pub","biergarten","nightclub"]
    },
    "groceries": {
      "shop":["convenience","grocer","greengrocer","food","supermarket"]
    },
    "pharmacies": {
      "amenity":["pharmacy","chemist","dispensing"]
    },
    "hotels": {
      "amenity":["hotel"],
      "tourism":["hotel"]
    },
    "libraries": {
      "amenity":["library"]
    },
    "education": {
      "amenity":["university","kindergarten","preschool","college","school"],
      "building":["school"]
    },
    "parks": {
      "leisure":["park"],
      "amenity":["park"]
    }
  }
```

**Comment by RELNO** [comment]
@crisjf what's you take on https://github.com/CityScope/CS_cityscopeJS/issues/110#issuecomment-751257756? how would you go by grabbing LBCS/NAICS for a random OSM tag?

**Comment by crisjf** [comment]
The paper looks like a great starting point!
